Bug 1454441: Hanging "System Lock" when executing "flush table ... fo…

gl-sergei opened the pull request:
…r export"

ibuf_merge_space returns non-zero value for "volume to be merged"
even if there were no pages found in change buffer for given table
space.

Caller function rely on this value to know when to stop change buffer
merge for given table space.

As a result, FLUSH TABLE FOR EXPORT wait until change buffer becomes
empty.

ibuf_get_merge_pages can return 0 even if there are pages to merge
(*n_pages > 0). It was the reason for increment sum_sizes.

The fix is to increment "sum_sizes" only when pages were found to
merge (*n_pages > 0).

gl-sergei commented:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/902/

laurynas-biveinis commented:
```
Why not remove the ++sum_sizes unconditionally? By looking into
its callers (ibuf_merge_space / ibuf_merge for a particular space
/ ibuf_contract_in_background), this should fix the bug the same,
and shouldn't have any other side effects.

Or is there an actual reason for the bump? The source history is
unhelpful. It could have been introduced for apparent symmetry
with ibuf_merge_pages doing the same, where it looks that it could
be removed the same. (This does not have to be a part of the
current fix).

Is it possible to add a MTR testcase? innodb-wl5522-debug does
something similar, it can be reviewed how to ensure that change
buffer is guaranteed to have some entries.
```

gl-sergei commented:
Callers of `ibuf_merge_space` invoke it again if there were some records
in `IBUF` to merge. `ibuf_merge_space` indicates this case via return
value (which is estimation of volume taken by `IBUF` records when they
are merged). However, in case of `DELETE` records, volume will be zero
(see `ibuf_rec_get_volume_func`), but merge still has to be done.
`++sum_sizes` just indicates that work has to be done.

I saw `innodb-wl5522-debug.test` and it was useful to catch a bug in my
initial fix (I placed wrong debug assertion failure checking that
`sum_sizes` equals 0 if and only if `*n_pages` equals zero, but this is
not the case). It might be possible to construct the test case, but it
will not be trivial. I'll spend some time on the test case.

gl-sergei commented:
I mean if we just remove `++sum_sizes`, then I can imagine a case when insert buffer will not be merged for exported table. When there are a lot of buffered deletes for that table for example.

laurynas-biveinis commented:
I see, thanks, the code change is OK then.

gl-sergei commented:
Test case added. I also disabled background statistics calculation when ibuf_disable_background_merge is set, because it caused change buffer merges.

http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/905/

laurynas-biveinis commented:
The Jenkins job does not show the added test being run, because it's marked as a big test, and no --big-test was passed to MTR:

00:07:03.774 main.percona_bug1454441                  [ skipped ]  Test needs 'big-test' option

Is there an actual reason for the test being marked as "big", what is its run time? Also, the test should probably go to the "innodb" suite, shouldn't it?

gl-sergei commented:
1. Test case isn't "big" anymore (actually, it was big initially, but then I reduced the size of t1 and amount of changes)
2. Moved to "innodb" suite
3. Fixed server crash caused by test case - missed `SET GLOBAL innodb_purge_run_now=ON;` (MTR didn't report that crash as server failure)
4. Removed unnecessary include 'have_debug_sync' and `set debug=` statement
   Jenkins run with `--big-test`, because some of IBUF tests require `--big-test`:
   http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/906/
